File: Source/Aladdin.Net/SerialException.cs
using System;
using System.Security; 
using System.Security.Permissions; 
using System.Runtime.Serialization; 

namespace Aladdin.Net
{
    ///////////////////////////////////////////////////////////////////////////
    // Сериализуемое исключение
    ///////////////////////////////////////////////////////////////////////////
    [Serializable]
    public class SerialException : ApplicationException
    {
#if (NET40_OR_GREATER || NETSTANDARD || NETCOREAPP)
        // состояние исключения
        [NonSerialized] private State state = new State();
        
        // конструктор
        private SerialException(string message, string stackTrace) : base(message)
        { 
            // сохранить переданные параметры            
            state.StackTrace = stackTrace; 

            // указать способ сохранения объекта
            SerializeObjectState += delegate(
                object exception, SafeSerializationEventArgs eventArgs)
            {
                // указать состояние объекта
                eventArgs.AddSerializedState(state);
            };
        } 
        // стековый фрейм исключения
        public override string StackTrace { get { return state.StackTrace; }}

        [Serializable]
        private struct State : ISafeSerializationData
        {
            // стековый фрейм исключения
            public string StackTrace; 

            // завершить десериализацию
            public void CompleteDeserialization(object obj)
            {
                // восстановить состояние
                ((SerialException)obj).state = this;
            }
        }
#else
        // конструктор
        protected SerialException(SerializationInfo info, StreamingContext context)

            // выполнить десериализацию
            : base(info, context) { stackTrace = info.GetString("stackTrace"); }

        // конструктор
        private SerialException(string message, string stackTrace) : base(message)

            // сохранить переданные параметры            
            { this.stackTrace = stackTrace; } private string stackTrace;

        // стековый фрейм исключения
        public override string StackTrace { get { return stackTrace; } }

        // сериализовать данные
        [SecuritySafeCritical]
        [SecurityPermission(SecurityAction.LinkDemand, SerializationFormatter = true)]
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            // сохранить стековый фрейм исключения
            base.GetObjectData(info, context); info.AddValue("stackTrace", stackTrace);
        }
#endif 
        // закодировать исключение
        public static string ToString(Exception exception, bool escape)
        {
            // получить описание ошибки и стековый фрейм
            string message = exception.Message; string stackTrace = exception.StackTrace; 

            // объединить описание ошибки и стековый фрейм
            if (!escape) return String.Format("{0}\0{1}", message, stackTrace); 

            // заменить символы перевода строки
            message    = message   .Replace(Environment.NewLine, "\n"); 
            stackTrace = stackTrace.Replace(Environment.NewLine, "\n"); 

            // выполнить экранирование
            message    = message   .Replace("\\", "\\\\").Replace("\n", "\\n");
            stackTrace = stackTrace.Replace("\\", "\\\\").Replace("\n", "\\n"); 

            // объединить описание ошибки и стековый фрейм
            return String.Format("{0}\n{1}", message, stackTrace); 
        }
        // раскодировать исключение
        public static Exception FromString(string description)
        {
            // найти разделитель строк
            int index = description.IndexOf('\0'); if (index >= 0)
            {
                // извлечь описание ошибки
                string message = description.Substring(0, index); 

                // извлечь стековый фрейм
                string stackTrace = description.Substring(index + 1); 

                // вернуть исключение
                return new SerialException(message, stackTrace);
            }
            // найти разделитель строк
            index = description.IndexOf('\n'); if (index >= 0)
            {
                // извлечь описание ошибки
                string message = description.Substring(0, index); 

                // извлечь стековый фрейм
                string stackTrace = description.Substring(index + 1); 
                
                // отменить экранирование
                message    = message   .Replace("\\n", "\n").Replace("\\\\", "\\");
                stackTrace = stackTrace.Replace("\\n", "\n").Replace("\\\\", "\\");

                // изменить символы перевода строки
                message    = message   .Replace("\n", Environment.NewLine); 
                stackTrace = stackTrace.Replace("\n", Environment.NewLine); 

                // вернуть исключение
                return new SerialException(message, stackTrace);
            }
            // вернуть исключение
            return new SerialException(description, String.Empty);
        }
    }
}

File: Source/Aladdin.Net.Java/src/aladdin/net/SerialException.java
package aladdin.net;
import aladdin.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////
// Исключение
///////////////////////////////////////////////////////////////////////////
public class SerialException extends IOException
{
    // номер версии при сериализации
    private static final long serialVersionUID = -1896353424275635162L;
    
    // закодировать исключение
    public static String toString(Throwable exception, boolean escape)
    {
        // получить описание исключения
        String message = exception.toString(); 
        
        // получить стековый фрейм
        String[] stackTrace = StackTrace.fromException(exception); 
        
        // создать строковый буфер 
        String str = StackTrace.toString(stackTrace); 

        // объединить описание ошибки и стековый фрейм
        if (!escape) return String.format("%1$s\000%2$s", message, str); 

        // получить разделитель строк
        String nl = System.getProperty("line.separator"); 
        
        // заменить символы перевода строки
        message = message.replace(nl, "\n"); 
        str     = str    .replace(nl, "\n"); 

        // выполнить экранирование
        message = message.replace("\\", "\\\\").replace("\n", "\\n"); 
        str     = str    .replace("\\", "\\\\").replace("\n", "\\n"); 

        // объединить описание и стек исключения
        return String.format("%1$s\n%2$s", message, str); 
    }
    // раскодировать исключение
    public static Exception fromString(String description)
    {
        // найти разделитель строк
        int index = description.indexOf('\000'); if (index >= 0)
        {
            // извлечь описание ошибки
            String message = description.substring(0, index); 

            // извлечь стековый фрейм
            String stackTrace = description.substring(index + 1); 

            // вернуть исключение
            return new SerialException(message, stackTrace);
        }
        // найти разделитель строк
        index = description.indexOf('\n'); if (index >= 0)
        {
            // извлечь описание ошибки
            String message = description.substring(0, index); 

            // извлечь стековый фрейм
            String stackTrace = description.substring(index + 1); 

            // отменить экранирование
            message    = message   .replace("\\n", "\n").replace("\\\\", "\\");
            stackTrace = stackTrace.replace("\\n", "\n").replace("\\\\", "\\");

            // получить разделитель строк
            String nl = System.getProperty("line.separator"); 
            
            // изменить символы перевода строки
            message    = message   .replace("\n", nl); 
            stackTrace = stackTrace.replace("\n", nl); 

            // вернуть исключение
            return new SerialException(message, stackTrace);
        }
        // вернуть исключение
        return new SerialException(description, new String());
    }
    // стековый фрейм исключения
    private final String stackTrace;
        
    // конструктор
    private SerialException(String message, String stackTrace) { super(message); 
            
        // получить разделитель строк
        String nl = System.getProperty("line.separator"); if (stackTrace.endsWith(nl)) 
        {
            // удалить завершающий перевод строки 
            stackTrace = stackTrace.substring(0, stackTrace.length() - nl.length()); 
        }
        // проверить наличие данных 
        if (stackTrace.length() == 0) this.stackTrace = message; 
        else { 
            // выполнить замену
            stackTrace = nl + "\tat " + stackTrace.replace(nl, nl + "\tat "); 
    
            // добавить описание исключения
            this.stackTrace = message + stackTrace; 
        }
    }
    // стековый фрейм исключения
    @Override public void printStackTrace() { printStackTrace(System.err); }
    
    // стековый фрейм исключения
    @Override public void printStackTrace(PrintStream s) { s.print(stackTrace); }
    @Override public void printStackTrace(PrintWriter s) { s.print(stackTrace); }
}
